cli/src/index.ts

```diff
@@ -15,63 +23,249 @@ export type RootCommand = typeof rootCommand;
+  withCommonOptions(command)
+    .group([jiraConsumerKeyOptionKey, jiraConsumerSecretOptionKey], "Auth:")
+    .demandOption([jiraHostOptionKey, jiraConsumerSecretOptionKey]);
 
 // eslint-disable-next-line @typescript-eslint/explicit-function-return-type
 export const withAuthOptions = <C extends RootCommand>(command: C) =>
   withCommonOptions(command)
-    .option("accessToken", {
+    .option(jiraAccessTokenOptionKey, {
       alias: "t",
       type: "string",
       describe: "access token",
+      default: config?.accessToken,
     })
-    .option("accessSecret", {
+    .option(jiraAccessSecretOptionKey, {
       alias: "s",
       type: "string",
       describe: "access secret",
+      default: config?.accessSecret,
+    })
+    .option(jiraUsernameOptionKey, {
+      alias: "u",
+      type: "string",
+      describe: "your jira username",
+      default: config?.username,
+    })
+    .option(jiraPasswordOptionKey, {
+      alias: "z",
+      type: "string",
+      describe: "your jira password",
+      default: config?.password,
+    })
+    .option(jiraPersonalAccessTokenOptionKey, {
+      alias: "k",
+      type: "string",
+      describe: "your jira personal access token",
+      default: config?.personalAccessToken,
+    })
+    .group(
+      [
+        jiraConsumerKeyOptionKey,
+        jiraConsumerSecretOptionKey,
+        jiraAccessTokenOptionKey,
+        jiraAccessSecretOptionKey,
+      ],
+      "Connection - OAuth:"
+    )
+    .group(
+      [jiraUsernameOptionKey, jiraPasswordOptionKey],
+      "Connection - User Credentials:"
+    )
+    .group(
+      [jiraPersonalAccessTokenOptionKey],
+      "Connection - Personal Access Token:"
+    )
+    .option("jira", {
+      describe: "Provide at least one full set of connection parameters",
     })
-    .group(["accessToken", "accessSecret"], authOptionGroup)
-    .demandOption(["accessToken", "accessSecret"]);
+    .coerce(
+      "jira",
+      (jira: {
+        readonly protocol?: "http" | "https";
+        readonly host?: string;
+        readonly consumerKey?: string;
+        readonly consumerSecret?: string;
+        readonly accessToken?: string;
+        readonly accessSecret?: string;
+        readonly personalAccessToken?: string;
+        readonly username?: string;
+        readonly password?: string;
+      }): JiraClient => {
+        // eslint-disable-next-line functional/no-conditional-statement
+        if (jira.protocol === undefined) {
+          // eslint-disable-next-line functional/no-throw-statement
+          throw new Error(
+            "Please provide the [jira.protocol] for accessing Jira"
+          );
+        }
+        // eslint-disable-next-line functional/no-conditional-statement
+        if (jira.host === undefined) {
+          // eslint-disable-next-line functional/no-throw-statement
+          throw new Error("Please provide the [jira.host] for accessing Jira");
```

Comment:
If we must check, we should accumulate errors and tell the user *everything* missing in one step.

Reply:
Good point. I'll fix that.

Reply:
I've done the best I can here with yargs. It doesn't really allow you to add your errors to its own checks. It runs your check first and if that fails it prints those errors. If it passes, it runs the yargs checks. So now you add your credentials to connect and if anything is missing, it'll tell you what you need to add. It will even try to guess how you're trying to connect, with OAuth being the default. Once you're connection parameters are all correct, it will do the rest of the parameter checks. So if I run it without the accessSecret and consumerKey I see:

```sh
➜  jiralint git:(feat/connection-options) ✗ cli/dist/jiralint search -t token -h jira.agiledigital.com.au
jiralint search

searches for jira issues using JQL and then lints

Common Required:
  -h, --jira.host      The Jira host                                    [string]
  -p, --jira.protocol  The Jira protocol
                          [string] [choices: "http", "https"] [default: "https"]

Connection - OAuth:
      --jira.consumerKey, --ck     The Jira consumer key
                                               [string] [default: "jiralintkey"]
      --jira.consumerSecret, --cs  The Jira consumer secret in base64 format
                                                                        [string]
  -t, --jira.accessToken           access token                         [string]
  -s, --jira.accessSecret          access secret                        [string]

Connection - User Credentials:
  -u, --jira.username  your jira username                               [string]
  -z, --jira.password  your jira password                               [string]

Connection - Personal Access Token:
  -k, --jira.personalAccessToken  your jira personal access token       [string]

Options:
      --version                 Show version number                    [boolean]
      --help                    Show help                              [boolean]
      --jira                    Provide at least one full set of connection
                                parameters                            [required]
      --qualityFieldName        The name of the Jira custom field used to store
                                issue quality                [string] [required]
      --qualityReasonFieldName  The name of the Jira custom field used to store
                                issue quality reason         [string] [required]
  -j, --jql                     jql to search by             [string] [required]
  -o, --output                  output format for results
                                   [choices: "json", "table"] [default: "table"]
      --boardNamesToIgnore      Prefix of the name of boards to be ignored when
                                determining the 'column' that a ticket is
                                currently in.              [array] [default: []]
      --customFieldNames        List of other custom issue field names to
                                include when retrieving issues from Jira.
                                                           [array] [default: []]

Missing required arguments:
  jira.consumerSecret
  jira.accessSecret
```

Once I add these, I see the quality fields and jql parameters are missing:

```sh
➜  jiralint git:(feat/connection-options) ✗ cli/dist/jiralint search -t token -h jira.agiledigital.com.au -s secret --cs consumersecret
jiralint search

searches for jira issues using JQL and then lints

Common Required:
  -h, --jira.host      The Jira host                                    [string]
  -p, --jira.protocol  The Jira protocol
                          [string] [choices: "http", "https"] [default: "https"]

Connection - OAuth:
      --jira.consumerKey, --ck     The Jira consumer key
                                               [string] [default: "jiralintkey"]
      --jira.consumerSecret, --cs  The Jira consumer secret in base64 format
                                                                        [string]
  -t, --jira.accessToken           access token                         [string]
  -s, --jira.accessSecret          access secret                        [string]

Connection - User Credentials:
  -u, --jira.username  your jira username                               [string]
  -z, --jira.password  your jira password                               [string]

Connection - Personal Access Token:
  -k, --jira.personalAccessToken  your jira personal access token       [string]

Options:
      --version                 Show version number                    [boolean]
      --help                    Show help                              [boolean]
      --jira                    Provide at least one full set of connection
                                parameters                            [required]
      --qualityFieldName        The name of the Jira custom field used to store
                                issue quality                [string] [required]
      --qualityReasonFieldName  The name of the Jira custom field used to store
                                issue quality reason         [string] [required]
  -j, --jql                     jql to search by             [string] [required]
  -o, --output                  output format for results
                                   [choices: "json", "table"] [default: "table"]
      --boardNamesToIgnore      Prefix of the name of boards to be ignored when
                                determining the 'column' that a ticket is
                                currently in.              [array] [default: []]
      --customFieldNames        List of other custom issue field names to
                                include when retrieving issues from Jira.
                                                           [array] [default: []]

Missing required arguments: qualityFieldName, qualityReasonFieldName, jql
```